lib/topics/index.ts

```diff
@@ -0,0 +1,177 @@
+import * as Promise from 'bluebird';
+import * as _ from 'lodash';
+import { HandlerFn, IRequestOptions, ISubscription, Mediator } from '../mediator';
+
+export interface IErrorWithDataId extends Error {
+  /** Identifier to use as a suffix on the 'error:' topic */
+  id?: string;
+}
+
+interface IHasId {
+  id?: string;
+}
+
+export default class Topics {
+  public mediator: Mediator;
+  public prefix: string;
+  public entity: string;
+  private subscriptions: {
+    [key: string]: ISubscription;
+  };
+
+  constructor (mediator: Mediator) {
+    this.mediator = mediator;
+    this.subscriptions = {};
+  }
+
+  /**
+   * Sets the prefix configuration for this instance, which will be part of the name
+   * of the handled topics.
+   * @param  {String} prefix
+   * @return {Topics}        returns self for chaining
+   */
+  public withPrefix(prefix: string) {
```

Comment:
I like inference for return types, but beware of one issue I've run into a few times: I'm occasionally careless and don't actually look at the inferred return type. Occasionally I return the wrong thing in one branch, and then I end up with a union type that I didn't expect. Your tests/use cases should hopefully point out the issue, but but if someone else is the only one consuming the function, they may not realize that the union type was actually a mistake.